gax-java/gax/src/test/java/com/google/api/gax/retrying/AbstractRetryingExecutorTest.java

```diff
@@ -222,29 +227,27 @@ public void testTotalTimeoutExceeded() throws Exception {
   }
 
   @Test
-  public void testCancelOuterFutureBeforeStart() throws Exception {
-    FailingCallable callable = new FailingCallable(4, "request", "SUCCESS", tracer);
+  public void testCancelOuterFutureBeforeStart() {
+    FailingCallable callable = new FailingCallable(0, "request", "FAILURE", tracer);
 
     RetrySettings retrySettings =
         FAST_RETRY_SETTINGS
             .toBuilder()
             .setInitialRetryDelay(Duration.ofMillis(1_000L))
             .setMaxRetryDelay(Duration.ofMillis(1_000L))
-            .setTotalTimeout(Duration.ofMillis(10_0000L))
+            .setTotalTimeout(Duration.ofMillis(10_000L))
             .build();
     RetryingExecutorWithContext<String> executor =
         getExecutor(getAlgorithm(retrySettings, 0, null));
     RetryingFuture<String> future = executor.createFuture(callable, retryingContext);
+    callable.setExternalFuture(future);
     boolean res = future.cancel(false);
-
+    verifyNoMoreInteractions(tracer);
```

Comment:
Can we still verify it as the last step like previously? Or it has to be verified earlier now?

Reply:
I think it makes sense to verify it here (before the executor submits the future) to ensure that the callable's logic isn't invoked (not invoking == exiting once it realizes the external future is done) after it's been submitted for execution.

Otherwise we can use verifyNoInteractions(tracer)  and put that at the end which requires an updating mockito version in dependencies.properties. Functionality-wise, they should be the same so I just used `verifyNoMoreInteractions` which already exists

Reply:
Are they the same functional-wise? The callable would never run until `executor.submit(future)`, and I think the purpose of this test is to verify that the inner callable would never run if the future is already cancelled, despite that the future is submitted. So verifying after `executor.submit(future)` makes more sense to me.

Reply:
The inner callable does run, but does not execute any logic -- `call()` is invoked by then immediately exited as the external future is always done. Since this is the first attempt, the `random retry delay` is set to 0 and is scheduled to execute immediately. I was thinking that this would create a possible race between when the callable exits and when  we run `verifyNoMoreInteractions`.

If `call()` runs before `verifyNoMoreInteractions` then there wouldn't be a point of running `verifyNoMoreInteractions` as the callable has already exited
If `call()` runs after `verifyNoMoreInteractions`, then this line would confirm the behavior. 

(From my understanding) We can't guarantee when the scheduleExecutor executes the callable (either before or after `verifyNoMoreInteractions` in the test) so asserting `verifyNoMoreInteractions` before the execution would confirm that the `call()` ran and that the tracer never ran.

Reply:
Which line would invoke `call()` immediately? I still think the callable is not invoked until `executor.submit(future)` based on the [Javadoc](https://github.com/googleapis/gapic-generator-java/blob/59efbbbd485282e37abbc5eed87b72a15a33240e/gax-java/gax/src/main/java/com/google/api/gax/retrying/RetryingExecutor.java#L67)

Reply:
Er* It *should* be invoked immediately once `executor.submit()` runs:
https://github.com/googleapis/gapic-generator-java/blob/59efbbbd485282e37abbc5eed87b72a15a33240e/gax-java/gax/src/main/java/com/google/api/gax/retrying/ScheduledRetryingExecutor.java#L115-L119 as `retryingFuture.getAttemptSettings().getRandomizedRetryDelay().toMillis()` -> 0ms for the first attempt

I looked up the `verifyNoMoreInteractions` javadoc and I misunderstood the use case. I assumed it was a point-in-time verification to ensure there was no more interactions once it was invoked (the race logic I mentioned above), but  it's meant to ensure that there were no more interactions after we've verified that things have already run/ executed/ invoked a certain number of times.

I think having either `verifyNoInteractions` or `verifyNoMoreInteractions` at the end of the file *should* work. May make more sense to have this be `verifyNoInteractions` instead of `verifyNoMoreInteractions` for clarity (tracer should have 0 interactions and `no more` implies more than 0) but that would require Mockito 4.x in the `dependencies.properties` file.